OpenRA.Mods.Common/Traits/BotModules/Squads/States/StateBase.cs

```diff
@@ -150,5 +150,73 @@ protected static bool ReloadsAutomatically(IEnumerable<AmmoPool> ammoPools, Rear
 
 			return true;
 		}
+
+		// Retreat units from combat, or for supply only in idle
+		protected void Retreat(Squad squad, bool flee, bool rearm, bool repair)
+		{
+			var loc = new CPos(0, 0, 0);
+
+			// HACK: "alreadyRepair" is to solve AI repair orders performance,
+			// which is only allow one goes to repairpad at the same time to avoid queueing too many orders.
+			// if repairpad logic is better we can just drop it.
+			var alreadyRepair = false;
+
+			if (flee)
+				loc = RandomBuildingLocation(squad);
+
+			foreach (var a in squad.Units)
+			{
+				if (IsRearming(a))
+					continue;
+
+				var orderQueued = false;
+
+				// Try rearm units.
+				if (rearm)
+				{
+					var ammoPools = a.TraitsImplementing<AmmoPool>().ToArray();
+					if (!ReloadsAutomatically(ammoPools, a.TraitOrDefault<Rearmable>()) && !FullAmmo(ammoPools))
+					{
+						squad.Bot.QueueOrder(new Order("ReturnToBase", a, orderQueued));
+						orderQueued = true;
```

Comment:
You may want to `continue` so the other orders don't add conflicts.

Reply:
In this place, because load ammo order can be queue with repair order, so I came up with this idea. only flee order is unqueueable.

Reply:
To put it more clearly:
If you play this branch you will find Aircraft can repair itself relatively continuously, while vehicle cannot. It is due to air squad state won't release all units after reaching fleet state but ground state will, which means for units in ground state the "Retreat" will trigger only once. The queued orders here are to make sure a vehicle can restore ammo as well as health even with this single order.

Reply:
meanwhile, queued orders on repair and reload will make aircraft queue too many orders (air squad will trigger `Retreat` many times)when they are waiting for repairing which will be hidden performance threats, so I only allow first damaged unit in squad to go repairing. This idea make Ground squad works not that ideal but compatible with air squad performance.

Reply:
The main problem is the `Retreat` needs using in different states and compatible for all to avoid massive duplication.